hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationMarkerChore.java

```diff
@@ -0,0 +1,132 @@
+ */
+@InterfaceAudience.Private
+public class ReplicationMarkerChore extends ScheduledChore {
+  private static final Logger LOG = LoggerFactory.getLogger(ReplicationMarkerChore.class);
+  private static final MultiVersionConcurrencyControl MVCC = new MultiVersionConcurrencyControl();
+  public static final RegionInfo REGION_INFO =
+    RegionInfoBuilder.newBuilder(REPLICATION_SINK_TRACKER_TABLE_NAME).build();
+  private static final String DELIMITER = "_";
+  private final Configuration conf;
+  private final RegionServerServices rsServices;
+  private WAL wal;
+  Random random = new Random();
+
+  public static final String REPLICATION_MARKER_ENABLED_KEY =
+    "hbase.regionserver.replication.marker.enabled";
+  public static final boolean REPLICATION_MARKER_ENABLED_DEFAULT = false;
+
+  public static final String REPLICATION_MARKER_CHORE_DURATION_KEY =
+    "hbase.regionserver.replication.marker.chore.duration";
+  public static final int REPLICATION_MARKER_CHORE_DURATION_DEFAULT = 30 * 1000; // 30 seconds
+
+  public ReplicationMarkerChore(final Stoppable stopper, final RegionServerServices rsServices,
+    int period, Configuration conf) {
+    super("ReplicationTrackerChore", stopper, period);
+    this.conf = conf;
+    this.rsServices = rsServices;
+  }
+
+  @Override
+  protected void chore() {
+    if (wal == null) {
+      try {
+        wal = rsServices.getWAL(null);
+      } catch (IOException ioe) {
+        LOG.warn("Unable to get WAL ", ioe);
+        // Shouldn't happen. Ignore and wait for the next chore run.
+        return;
+      }
+    }
+    String serverName = rsServices.getServerName().getServerName();
+    long timeStamp = EnvironmentEdgeManager.currentTime();
+    // We only have timestamp in ReplicationMarkerDescriptor and the remaining properties walname,
+    // regionserver name and wal offset at ReplicationSourceWALReaderThread.
+    byte[] rowKey = getRowKey(serverName, timeStamp);
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Creating replication marker edit.");
+    }
+
+    // This creates a new ArrayList of all the online regions for every call.
+    List<HRegion> regions = rsServices.getRegions();
+
+    if (regions.isEmpty()) {
+      LOG.info("There are no online regions for this server, so skipping adding replication marker"
+        + " rows for this regionserver");
+      return;
+    }
+    HRegion region = regions.get(random.nextInt(regions.size()));
+    try {
+      WALUtil.writeReplicationMarkerAndSync(wal, MVCC, region.getRegionInfo(), rowKey, timeStamp);
+    } catch (IOException ioe) {
+      LOG.error("Exception while sync'ing replication tracker edit", ioe);
+      // TODO: Should we stop region server or add a metric and keep going.
+    }
+  }
+
+  /**
+   * Creates a rowkey with region server name and timestamp.
+   * @param serverName region server name
+   * @param timestamp  timestamp n
```

Comment:
Delete the last n? It should a bug of our old spotless rule...

Reply:
Removed in latest commit